sled-agent/src/support_bundle/storage.rs

```diff
@@ -0,0 +1,1393 @@
+            };
+
+            let bundle = SupportBundleMetadata { support_bundle_id, state };
+            bundles.push(bundle);
+        }
+
+        Ok(bundles)
+    }
+
+    /// Returns the hex, lowercase sha2 checksum of a file at `path`.
+    async fn sha2_checksum_matches(
+        path: &Utf8Path,
+        expected: &ArtifactHash,
+    ) -> Result<bool, Error> {
+        let mut buf = vec![0u8; 65536];
+        let mut file = tokio::fs::File::open(path).await?;
+        let mut ctx = sha2::Sha256::new();
+        loop {
+            let n = file.read(&mut buf).await?;
+            if n == 0 {
+                break;
+            }
+            ctx.write_all(&buf[0..n])?;
+        }
+
+        let digest = ctx.finalize();
+        return Ok(digest.as_slice() == expected.as_ref());
+    }
+
+    // A helper function which streams the contents of a bundle to a file.
+    //
+    // If at any point this function fails, the temporary file still exists,
+    // and should be removed.
+    async fn write_and_finalize_bundle(
+        mut tmp_file: tokio::fs::File,
+        from: &Utf8Path,
+        to: &Utf8Path,
+        expected_hash: ArtifactHash,
+        stream: impl Stream<Item = Result<Bytes, HttpError>>,
+    ) -> Result<(), Error> {
+        futures::pin_mut!(stream);
+
+        // Write the body to the file
+        let mut hasher = Sha256::new();
+        while let Some(chunk) = stream.next().await {
+            let chunk = chunk?;
+            hasher.update(&chunk);
+            tmp_file.write_all(&chunk).await?;
+        }
+        let digest = hasher.finalize();
+        if digest.as_slice() != expected_hash.as_ref() {
+            return Err(Error::HashMismatch);
+        }
+
+        // Rename the file to indicate it's ready
+        tokio::fs::rename(from, to).await?;
+        Ok(())
+    }
+
+    /// Creates a new support bundle on a dataset.
+    pub async fn create(
+        &self,
+        zpool_id: ZpoolUuid,
+        dataset_id: DatasetUuid,
+        support_bundle_id: SupportBundleUuid,
+        expected_hash: ArtifactHash,
+        stream: impl Stream<Item = Result<Bytes, HttpError>>,
+    ) -> Result<SupportBundleMetadata, Error> {
+        let log = self.log.new(o!(
+            "operation" => "support_bundle_create",
+            "zpool_id" => zpool_id.to_string(),
+            "dataset_id" => dataset_id.to_string(),
+            "bundle_id" => support_bundle_id.to_string(),
+        ));
+        let root =
+            self.get_configured_dataset(zpool_id, dataset_id).await?.name;
+        let dataset =
+            NestedDatasetLocation { path: support_bundle_id.to_string(), root };
+        // The mounted root of the support bundle dataset
+        let support_bundle_dir = dataset
+            .mountpoint(illumos_utils::zpool::ZPOOL_MOUNTPOINT_ROOT.into());
+        let support_bundle_path = support_bundle_dir.join("bundle");
+        let support_bundle_path_tmp = support_bundle_dir.join("bundle.tmp");
```

Comment:
I think I hadn't considered this as a problem because the Nexus writing the support bundle should have unique ownership off it -- support bundles are assigned to an individual Nexus for creation and storage -- but regardless, I can use a randomized path.